src/Functions/FunctionsStringDistance.cpp

```diff
@@ -271,6 +272,222 @@ struct ByteEditDistanceImpl
+        std::fill_n(s1_matching, s1len, -1);
+        std::fill_n(s2_matching, s2len, -1);
+
+        // calculate matching characters
+        int matching_characters = 0;
+        for (int i = 0; i < s1len; i++) {
+            // matching window
+            int min_index = std::max(i - max_range, 0);
+            int max_index = std::min(i + max_range + 1, s2len);
+            if (min_index >= max_index) break;
+
+            for (int j = min_index; j < max_index; j++) {
+              if (s2_matching[j] == -1 && haystack[i] == needle[j]) {
+                s1_matching[i] = i;
+                s2_matching[j] = j;
+                matching_characters++;
+                break;
+              }
+            }
+        }
+
+        if (matching_characters == 0) {
+            delete [] s1_matching;
+            delete [] s2_matching;
+            return ResultType(0.0);
+        }
+
+        // transpositions (one-way only)
+        double transpositions = 0.0;
+        for (int i = 0, s1i = 0, s2i = 0; i < matching_characters; i++) {
+        while (s1_matching[s1i] == -1) {
+          s1i++;
+        }
+        while (s2_matching[s2i] == -1) {
+          s2i++;
+        }
+        if (haystack[s1i] != needle[s2i]) transpositions += 0.5;
+            s1i++;
+            s2i++;
+        }
+        double m = static_cast<double>(matching_characters);
+        double jaro_similarity = 1.0 / 3.0  * ( m / static_cast<double>(s1len)
+                                            + m / static_cast<double>(s2len)
+                                            + (m - transpositions) / m );
+        delete [] s1_matching;
+        delete [] s2_matching;
+        return ResultType(jaro_similarity);
+    }
+};
+
+// Based on https://en.wikipedia.org/wiki/Jaro%E2%80%93Winkler_distance
+// Implements Jaro similarity
+struct ByteJaroWinklerSimilarityImpl {
+
+    using ResultType = Float64;
+    static constexpr size_t max_string_size = 1u << 16;
+
+    static ResultType inline process(
+        const char * __restrict haystack, size_t haystack_size, const char * __restrict needle, size_t needle_size)
+    {
+        constexpr int MAX_PREFIX_LENGTH = 4;
+        double scaling_factor =  0.10;
+        double boost_threshold = 0.7;
+
+        int s1len = static_cast<int>(haystack_size);
+        int s2len = static_cast<int>(needle_size);
+
+
+        /// Safety threshold against DoS, since we use two array to calculate the distance.
+        if (haystack_size > max_string_size || needle_size > max_string_size)
+            throw Exception(
+                ErrorCodes::TOO_LARGE_STRING_SIZE,
+                "The string size is too big for function editDistance, "
+                "should be at most {}", max_string_size);
+
+        ResultType jaro_similarity = ByteJaroSimilarityImpl::process(haystack, haystack_size, needle, needle_size);
+        if (jaro_similarity == -1.0) return ResultType(-1.0);
+
+        ResultType jaro_winkler_similarity = jaro_similarity;
```

Comment:
I find the use of both `jaro_similarity` and `jaro_winkler_similarity` a bit of inconvenient. I would do one of the followings:
1. Eliminate `jaro_similarity` and use `jaro_winkler_similarity` instead.
2. Eliminate `jaro_winkler_similarity` and directly return the values instead of storing them into a variable.